pkg/log/log.go

```diff
@@ -0,0 +1,222 @@
+	pflag.CommandLine.AddFlagSet(zapFlagSet)
+
+	// Add flags registered by imported packages (e.g. glog and
+	// controller-runtime)
+	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
+
+	pflag.Parse()
+
+	// Use a zap logr.Logger implementation. If none of the zap
+	// flags are configured (or if the zap flag set is not being
+	// used), this defaults to a production zap logger.
+	//
+	// The logger instantiated here can be changed to any logger
+	// implementing the logr.Logger interface. This logger will
+	// be propagated through the whole operator, generating
+	// uniform and structured logs.
+	logf.SetLogger(zap.Logger())
+
+	// set the logger.
+	log = Logger{
+		name:          withName,
+		lgr:           logf.Log.WithName(withName),
+		out:           os.Stdout,
+		isTestingMode: false,
+	}
+
+	return &log
+}
+
+func (p *Logger) SetOutput(out io.Writer, isTestingMode bool) *Logger {
+	// WithValues, WithName and ZapLoggerTo all result in a new logger instance.
+	// The values stored in the Logger struct must be set again.
+	if len(log.tags) > 0 {
+		log.lgr = logf.ZapLoggerTo(out, isTestingMode).WithName(log.name).WithValues(log.tags...)
+	} else {
+		log.lgr = logf.ZapLoggerTo(out, isTestingMode).WithName(log.name)
+	}
+
+	log.out = out
+	log.isTestingMode = isTestingMode
+	return &log
+}
+
+// GetLogger returns the current logger object.
+func GetLogger() *Logger {
+	return &log
+}
+
+// Info logs are used for non-error messages. It will log a message with
+// the given key/value pairs as context.
+func (p *Logger) Info(ctx *gin.Context, msg string) *Logger {
+	return p.Infof(ctx, msg)
+}
+
+// Infof logs are used for non-error messages. It will log a message with
+// the given key/value pairs as context.
+func (p *Logger) Infof(ctx *gin.Context, msg string, args ...string) *Logger {
+	ctxInfo := addContextInfo(ctx)
+	arguments := make([]interface{}, len(args))
+	for i, arg := range args {
+		arguments[i] = arg
+	}
+	if len(arguments) > 0 {
+		p.lgr.Info(fmt.Sprintf(msg, arguments...), ctxInfo...)
+	} else {
+		p.lgr.Info(msg, ctxInfo...)
+	}
+
+	return p
+}
+
+// Error logs are used for logging errors. It will log the error with the given
+// message and key/value pairs as context.
+func (p *Logger) Error(ctx *gin.Context, err error, msg string) *Logger {
+	return p.Errorf(ctx, err, msg)
+}
+
+// Errorf logs are used for logging errors. It will log the error with the given
+// message and key/value pairs as context.
+func (p *Logger) Errorf(ctx *gin.Context, err error, msg string, args ...string) *Logger {
+	ctxInfo := addContextInfo(ctx)
+	arguments := make([]interface{}, len(args))
+	for i, arg := range args {
+		arguments[i] = arg
+	}
+
+	if len(arguments) > 0 {
+		p.lgr.Error(err, fmt.Sprintf(msg, arguments...), ctxInfo...)
+	} else {
+		p.lgr.Error(err, msg, ctxInfo...)
+	}
+
+	return p
+}
+
+// WithValues appends tags to the logger.
+func (p *Logger) WithValues(keysAndValues ...interface{}) *Logger {
```

Comment:
As agreed we should use maps. Not slices.

Reply:
fixed: ab763014216a10160ac49c4b79f34ff2e6a77b1e